First four runtimes (#46)
val-town/val-town-blog: src/content/blog/first-four-val-town-runtimes.mdx

---
title: The first four Val Town runtimes
generated: 1701894028894
description: From vm to vm2 to workers to processes, the long journey to how we run vals today
pubDate: February 5, 2024
heroImage: ./first-four-val-town-runtimes/timeline.png
author: Tom MacWright
---

import { Image } from "astro:assets";
import memoryImage from "./first-four-val-town-runtimes/memory.png";

> One of the most important things that Val Town does is **run untrusted code**.

Reviewer: ```suggestion
The core of Val Town is **running untrusted code**.
```

> Our whole premise requires us to take the TypeScript code that our community
> writes and run it on a server, safely, quickly, and flexibly.

Reviewer: ```suggestion
We take your TypeScript code and run it on a server, safely, quickly, and flexibly.
```

Reviewer: You could call out that there are tradeoffs between all of these items, and figuring out how to balance them all, and get the most of each is the trick, and we keep iterating in that direction, blah blah blah (sorry, writing quick notes!)

#### Safety

The default way to run arbitrary code is "dangerously". Modules in Node.js have unfettered
access to your system's environment variables, filesystem, network, and much
more. This access is [routinely abused in supply-chain attacks](https://socket.dev/blog/inside-node-modules).
The easiest way to run an arbitrary string of code is with the `eval()` method,
which doesn't provide any kind of sandbox protections, so it can easily become

> an attack vector.

Reviewer 2: I understand the ideas being laid out in this paragraph, but they do feel disconnected

Reviewer: Agreed, could use more of a connection to what's above

Security generally means isolation and containment:

- Vals need to be isolated from the system they're running on: they shouldn't
  be able to read from the filesystem or learn about environment variables.
- Vals should be isolated from each other: you shouldn't be able to affect
  anyone else's val's behavior.

#### Speed

> On the other hand, if you were to make every Val run in a full-fledged Docker

Reviewer: You call out the tradeoff between safety and speed more explicitly

container, it'd be tremendously slow. You're booting up a whole Linux machine
with all the bells and whistles just to evaluate a few lines of code. It'd be
[pretty secure](https://docs.docker.com/engine/security/#conclusions), but
pretty slow.
You could create a pool of Docker containers, but the memory and devops overhead
would be significant.

> #### Flexibility

Reviewer: I might entitled this section "NPM" or "libraries" or "compatibility" or something like that

Val Town lets you bring a lot of experience from other systems. If you've used
Node.js, great - your NPM modules should mostly work. Thanks to our current system,
a lot of DOM methods work too - you can parse query strings with [URLSearchParams](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) and use new APIs
like web-standard streams. A way to square the safety + speed requirements
would be to use [QuickJS](https://bellard.org/quickjs/), a standalone JavaScript
engine that's fast to start up and easily sandboxed with WebAssembly. But
we'd be miles away from NPM or Node compatibility, and would run JavaScript
about 35x slower.

---

So, those are the properties we're aiming for. What have we tried so far? Buckle up.

#### Runtime 1: Node.js + vm

> The very first version of Val Town used Node.js’s [vm module](https://nodejs.org/api/vm.html).

Reviewer: You could caveat that this was a super early prototype that only a dozen or so folks used, and nobody used for sensitive or serious stuff

Reviewer 2: Yeah, when I joined in September, we were already using vm2

Author: 👍 made this clearer

Schematically, running a val looked like this:

```ts
const [returnValue, logs, error] = vm.runInContext(`
  let resolveValue;
  let returnValue;
  let error;
  const logs = [];
  console.log = (log) => { logs.push(log); return undefined };
  try {
    ${replacedCode}
    returnValue = ${name}
  } catch (e) {
    error = e;
  }
  [returnValue, logs, error];
`);
```

Now, the documentation for the vm module warns that "The node:vm module
is not a security mechanism. Do not use it to run untrusted code."
So it should be no surprise that this approach didn't last very long.

The Node.js vm module is incredibly convenient to use, but it was dangerous and
didn't check our boxes:

- Infinite loops in user code will take down your server
- It's easy to escape the sandbox and access sensitive environment variables
- You can't install NPM modules within the sandbox
- Scripts that run in the Node.js VM sandbox run in _script_ mode, not _module_
  mode, so you can't use `import` or `export`.

> #### Runtime 2: Node.js + vm2

Reviewer: Again, you could caveat that we moved off this a month after we launched, or something like that

  </figcaption>
</figure>

- However, Workers triggered [memory problems](https://github.com/denoland/deno/issues/18414)
  that made the evaluation server slower and slower over time until it restarted.
  We were spawning and killing a _lot_ of workers, and understandably this
  wasn't a typical usecase.
- Server isolation had a performance cost: for a val that powers a website, for each
  request we'd receive the request in the Express server, serialize it, send
  it to Deno, which would deserialize and operate on it, send the response to Express,
  which would then send it to the user.

Despite these issues, the switch to Deno was a big upgrade, and we were significantly
less affected by anything that untrusted code could do.

#### Runtime 4: Node.js + Deno processes

The current iteration of our runtime learns from the lessons so far:

- You can't really use JavaScript to build a JavaScript sandbox.
- Deno is a powerful tool for simple containment of functionality.
- Memory leaks are scary.

Right now, Vals run as temporary Deno subprocesses of a Node.js server.
The subprocess's code is tiny: it uses [node-deno-vm](https://github.com/casual-simulation/node-deno-vm),

> a terrific module from Kallyn Gowdy, which works by exposing a WebSocket

Reviewer 2: We should link this to his GitHub or website

```suggestion
a terrific module from [Kallyn Gowdy](https://github.com/KallynGowdy), which works by exposing a WebSocket
```

connection between Deno and Node.js and wrapping it in
a [postMessage](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage)-style API.
The "guest" code that wraps vals is also now [public](https://esm.town/guest.ts),
though it's still an implementation detail and could change at any time.

While this gives far better isolation between user code and system code
than our previous vm and Worker-based strategies, it takes a little more
time to start up, so we use a pool of pre-initialized processes to
avoid that time penalty.

This strategy also gives the ability to have more fine-grained communication
with Deno, because it's based on WebSockets instead of a simple REST
HTTP request cycle. We can send multiple messages to control vals while
they're running, and receive multiple messages from them as they run.

> This is key to how we just cut 100ms from typical val runtimes - more on
> that later.

Reviewer 2: > more on that later.

This made me think we'd talk about it later in this post. I'd be more explicit that this will be a future post.

```suggestion
This is key to how we just cut 100ms from typical val runtimes - more on
that on a future post.
```

| Isolation primitive | V8 Context | V8 Context   | V8 Context           | Process                     |
| Val syntax          | Custom JS  | Custom TS+JS | Custom TS+JS         | Standard TS+JS              |
| Importing vals      | `@` syntax | `@` syntax   | `@` syntax           | ESM-standard `import`       |

As you can tell, we've been iterating fast. v4 is not the last version of
the Val Town runtime: we still have more ground to cover on all of the dimensions that we
care about. It's heartening to see others contribute to this discussion - Figma's
adoption of QuickJS, after
[using Realms was not secure enough](https://www.figma.com/blog/an-update-on-plugin-security/),
Amazon open-sourcing [Firecracker](https://aws.amazon.com/blogs/aws/firecracker-lightweight-virtualization-for-serverless-computing/)
one of the components of Lambda, and Deno
sharing [details of how their Deploy product works](https://deno.com/blog/anatomy-isolate-cloud).

We want to learn from everyone. Vals should run instantly, securely, and everything should just
work.

Oh, and if you read this and you have the answer and the experience and are the

> one to unlock the next-gen runtime: we're hiring: [steve@val.town](mailto:steve@val.town).

Reviewer: This should link to our yet-to-be-posted hiring post. In theory we should wait until that's posted to post this.

Also: our call to action could also be: try val town. It's super fun and fast and flexible, like we said above!
